### PR TITLE
Prevent the loss of Field metadata when converting Fields into FieldDimensions

### LINKED ISSUE #18646: Cannot select granularity/binning on fields from SQL question, which is joined
**Describe the bug**
Chosing between year,quarter, month etc in Custom question group by time dimension is not possible when using joined table.

**Logs**
No log really just some UI issue.

**To Reproduce**
1. Create a custom question with a join
2. Add an aggregation
3. Search for a time dimension in the joined table field list
4. See that you cannot switch from day to month or any other level of detail. 

**Expected behavior**
I would expect to have the level of detail help in the field picker when grouping by a joined table field.

**Screenshots**

Screenshot 2021-10-25 at 16 10 25

**Information about your Metabase Installation:**
- Your browser and the version: Chrome/94.0.4606.81
- Your operating system: Mac OS X 10_15_7
- Your databases: Snowflake
- Metabase version: 1.41.0
- Metabase hosting environment: Hosted
- Metabase internal database: PostgreSQL

**Severity**
Not really blocking as we can invert tables but it's annoying and miss leading to anyone not very tech saavy.

**Additional context**

### PR DESCRIPTION
**Pull request by daltojohnso**
Fixes #18646

**Root cause**
1. In order to get a list of granularity/binning options for the dimensions found in a given query, we call `StructuredQuery.prototype.dimensionOptions`. This method collects these options from its joins via a call to `Join.prototype.joinedDimensionOptions`.
1. The Join instance's underlying table (coming from the `Join.prototype.joinedTable` method) is a virtual table in this scenario, because we're using a saved question in the join, with an id of something like `"card__123"`. This virtual table looks correct; it is retrievable via `metadata.table("card__123")` and it has a fully populated list of Fields with all metadata.
1. `joinedDimensionOptions` converts the fields found on the virtual table into `FieldDimension` instances. In this conversion process, `Field.prototype.dimension` returns the results from `Dimension.parseMBQL(this.reference(), this.metadata, this.query)`. The `this.query` value here is `undefined` because the field instance is treated as coming from a table (even if the table is actually a card, sort of), so when the Dimension is instantiated, all it has is the reference, eg something like `["field", "CREATED_AT", { "base-type": "type/DateTime" }]`.
1. When the above `FieldDimension` is converted back into a `Field` instance via `FieldDimension.prototype.field`, it does not have enough information to properly return it to its original form, so it hits the fallback at the end of the method and returns a `Field` with only an `id` and a `base-type`.
1. This is the point of no return -- we are very restricted in what we can do with this object. For granularity/binning to work, we rely on the presence of `dimension_options` on the field instance, and we no longer have way to access it.

**TLDR**: the Field --> FieldDimension --> Field transformation is lossy because we don't hold onto the field's `table_id`, which is needed to uniquely identify the fields on virtual tables.

In addition to the above, when creating the subdimensions for a given field, we are losing the `base-type` of the subdimension option, so the query will fail. One of the commits from https://github.com/metabase/metabase/pull/18921 fixes it: https://github.com/metabase/metabase/pull/21782/commits/d9170a4aae0f4bcae4aaca7755f6046eeb1ae02b

---
**Potential solution 1** 
1. When instantiating the `FieldDimension` from a field, preserve its `table_id` property.
1. Then, in `FieldDimension.prototype.field`, we can call `metadata.table(this.table_id)` to get the virtual table and then look for the field using `ref[1]` or `fieldIdOrName`.

**Potential solution 2**
1. When instantiating the `FieldDimension` directly from a `Field` instance set on the dimension instance a reference to the field instance.
1. Then, in `FieldDimension.prototype.field`, we can immediately check for the presence of this instance and return it.

---

**This PR uses solution 2. Here's why:**
1. The `FieldDimension.prototype.field` method **is a giant hack** and adding another hoop to jump through will probably end up hurting us in the long run. We potentially start running into scenarios where `table_id` is set on the dimension alongside a query. In that scenario, which `Field` do we use? Either/or? Or will there be a difference?
1. If we already have a reference to the `Field` instance, why can't we just hold onto it? It short circuits the giant hack field method, and it guarantees we preserve everything when we go from fields to dimensions to fields to dimensions.
1. We introduce some risk around relying on a bad field instance, but this is easily resolvable by identifying what fields can be trusted: fields found in the god `Metadata` object.

Regarding the addition of `_comesFromEndpoint`, this is obviously a hack, but I think it is a reasonable one to add, for now. We are willy nilly instantiating new `Field` instances in a number of places, and these client-side Field instances often do not contain all the field metadata. We should probably not be doing this to begin with, but for now, until we refactor this code, I think having some way of identifying _real_ instances will be useful.

---

**Testing**
1. Follow the steps in #18646

### CODE REVIEW

File: frontend/test/metabase/scenarios/joins/joins-2.cy.spec.js [hunk 1]
```diff
@@ -180,7 +180,7 @@ describe("scenarios > question > joined questions", () => {
 
       popover().within(() => {
         enterCustomColumnDetails({
-          formula: "[Question 5 → sum] / [Sum of Rating]",
+          formula: "[Question 5 → Sum of Rating] / [Sum of Rating]",
```
Comment (daltojohnso) [inline frontend/test/metabase/scenarios/joins/joins-2.cy.spec.js:183]: This patch fixes this bug where we were failing to find the joined field's display name in the expression editor. Is this an issue?

**Comment by daltojohnso** [comment]
FYI we depend on the field's fingerprint to determine which granularity option to show by default. Unfortunately, these virtual table fields seem to have a fingerprint that might either wrong or missing, so the initial option might not be "by month." Locally, I see "by minute"

Screen Shot 2022-04-20 at 1 43 17 PM

**Review by kulyk — APPROVED** [review]
Given how messed up `field` retrieval can be, this looks like the most elegant fix. Great work @daltojohnso 👏👏👏 
And kudos for explaining the problem + possible approaches in the PR description, that was an extremely helpful read!

**Comment by daltojohnso** [comment]
This change is simple enough imo that we can include it in the next release.
